### DISCUSSION TITLE
Better 1st party env handling

### DISCUSSION THREAD

**Discussion by florian-lefebvre** [withastro/roadmap#822] [Proposal]
### Body

# Summary

This RFC aims to improve DX (and eventually security) around env variables.

# Background & Motivation

I recently started working on https://github.com/florian-lefebvre/astro-env with pretty basic features, and it looked like it could be great to rework env handling as a whole in Astro core. Moreover, there was an issue about env recently (see https://github.com/withastro/astro/issues/9827). Other JS frameworks have better ways to handle env.

Env variables are an important part of any web application. You need to store sensitive data (think API secrets, tokens etc) without being leaked inside your git repo. But that's only the 1st part of the story. It's easy to leak this data by importing in the wrong place, eg. the frontend like Resend a few weeks ago.

For example, today it's hard:
- To validate the env vars
- To apply transforms to the env vars
- To type the env vars

# Goals

- Make env vars typing easier (eg. so that people don't have to do it manually in `env.d.ts`)
- Be able to validate env vars (eg. `SECRET_TOKEN` must be at least 64 long)
- Be able to transform env vars (eg. `PORT` is always a number)

# Example

## In other frameworks/tools

- https://nuxt.com/docs/guide/going-further/runtime-config
- https://kit.svelte.dev/docs/modules#$env-dynamic-private
- https://env.t3.gg/

## In Astro

Since Astro uses zod a lot, it could be great to also define the env schema using it. This could be using an entrypoint at `${rootDir}/env.{mjs|ts}` with a default export:

```ts
// env.ts
import { defineEnv } from "astro/env"
import { z } from "astro/zod"

export default defineEnv({
	schema: z.object({
		// ...
	})
})
```

This entrypoint would be validated inside a vite plugin (to warn/error on invalid env vars) and used inside a virtual module to expose env to the users. For example:

```ts
import { env } from "astro:env"

const port = env.PORT
```

Adopting virtual modules like SvelteKit could be useful to make clearer to users when they should or should not use non public variables.

---

I think it can be great to update `astro-env` according to the outcome of this RFC to test and gather feedback, and then merge it into core.

**Comment by lilnasy** [comment]
This is great work! Environment handling needs some attention so thank you for starting this discussion.

Some points of discussion I'm interested in:
- **Should we involve zod?** - Users really only care about whether or not a variable exists. Email, etc may be too niche to introduce a schema validator. I think this is worth considering because we have only used zod for build-time checks, we do not use it for SSR.
- **How should secret and public variables be distinguished?** - Following svelte's lead, we could create a `astro:env/public`. However, we already have a convention of requiring a prefix.
- **How should static and runtime variables be distinguished?** - We currently have awkward rules where...

1. astro-provided variables (MODE, PROD, BASE_URL, DEV, SITE, ASSETS_PREFIX) are always inlined

2. variables found in `.env` file are inlined

3. ...unless also available in `process.env`

4. only variables prefixed with `PUBLIC_` are inlined for client-side code

5. the rest are replaced with `process.env.*`, even for client-side code

**Reply by florian-lefebvre** [reply]
> Should we involve zod?

I think that can be powerful to have transforms and some more advanced checks like urls or unions. However, it may no be worth having the whole zod API at the user disposal. An idea could be to support a subset only (like https://github.com/florian-lefebvre/astro-env/pull/1/files#diff-037d15dcf3a9c89e66b49efacc634941743ba94a147dc1428f18c9814e8814d4R63-R68)

> How should secret and public variables be distinguished?

IIRC svelte exports public vars from `$env/static/public` and keeps the `PUBLIC_` prefix as well. I don't think that's a bad thing, it doesn't make the DX worse and it makes it easy to filter secret/public vars

> How should static and runtime variables be distinguished?

That's probably the most important part. Regarding 1., I think it could make sense to recommend accessing those using `import.meta.env` and any others using the virtual imports. As for the rest, I lack context to give any opinion

**Reply by alexanderniebuhr** [reply]
> Should we involve zod?

Personally I think we should let the user decide if they want to use zod or not. It should work both

> svelte exports public vars from $env/static/public

I dislike the idea of having to separate import paths, but I do agree that we should keep `PUBLIC_`

> We currently have awkward rules where...

Yes correct, and maybe not something this RFC has at it's goals, but we should definetly find a better set of rules. E.g. the once you listed are all valid, but I'm pretty sure adapters could introduce own urls e.g. the Cloudflare adapter has some additional rules.

**Comment by dreyfus92** [comment]
It's a significant step forward for Astro, and I appreciate the depth of thought you've put into this. I see the value in integrating it with Zod. It will add a layer of robustness and error-proofing that could be vital for complex projects. Perhaps making its use optional could bridge the gap between simplicity for basic use cases and depth for advanced ones.  What I think it's crucial is to establish an easy distinction between static and runtime variables and this pretty much addresses it. This seems to offer a clearer, more streamlined approach, which could be a big win for DX.

Lastly, the idea of using virtual modules for env vars, akin to SvelteKit, is innovative. It not only simplifies the process but also makes it more intuitive.

**Reply by florian-lefebvre** [reply]
> Perhaps making its use optional could bridge the gap between simplicity for basic use cases and depth for advanced ones

I like how it works for t3, see https://env.t3.gg/docs/customization#skipping-validation

> What I think it's crucial is to establish an easy distinction between static and runtime variables and this pretty much addresses it.

tbh if we follow something like sveltekit convention, that pretty much covers everything imo

**Reply by dreyfus92** [reply]
I would lean more towards the SvelteKit's convention tho.

**Comment by florian-lefebvre** [comment]
More thoughts: I think having the env defined in a separate file (entrypoint) is better for a few reasons:
- it does not make the astro config huge (it's easy to reach 20 env vars, and +20 lines in the config makes it harder to navigate)
- it's easier to use by code. Otherwise, it requires serializing things and if we choose to use zod, it's kind of a mess

However, exporting from an entrypoint called `env` has the downside of being possibly suggested by the IDE auto-completion. At least doing a default export seems safer than exporting a `env` object

**Reply by bholmesdev** [reply]
I understand this. But to be honest, I like giving the users a choice over introducing a new special file. One option could be keeping `defineEnv` and allowing users to split out like so:

```ts
// env.ts
export const env = defineEnv({
  schema: {...}
});

// astro.config.ts
import { env } from './env';

export default defineConfig({
  env, // assuming `env` is a top-level property of the Astro config
  integrations ...
});
```
I know some users are sensitive to new config files, so this solution would allow them to inline while allowing others to split out separately.

I also don't want to set a precedent for splitting more complex config to new file names. It's nice keeping our reserved file surface low!

**Comment by alexanderniebuhr** [comment]
I apologize in advance but I'm just posting some thoughts :re specifically Cloudflare here:

For Cloudflare we do have some special "rules":

https://github.com/t3-oss/t3-env/issues/60
https://github.com/t3-oss/t3-env/issues/65
https://github.com/withastro/astro/issues/7137

Just to give perspective, this is a `t3-env.ts`, I need to make it working as expected. 

```ts
import { createEnv } from '@t3-oss/env-core'
import { z, ZodError } from 'zod'

const isServer = typeof window === 'undefined'
const isBuild = isServer && process.env.STATE === 'building'

const serverVariables = {
	build: {
		// Put in here any server-side variable that is hardcoded on build
		VITE_TIME: z.string().nonempty(),
	},
	runtime: {
		// Put in here any server-side variable that is read at runtime
		CF_PAGES_URL: z.string().nonempty(),
		DATABASE_URL: z.string().nonempty(),
	},
}

export const createRuntimeEnv = (prop?: unknown) => {
	if (isBuild) {
		console.log(
			'Please make sure that following variables are set in your server runtime environment: ',
			Object.keys(serverVariables.runtime).join(', '),
		)
	}
	const rEnv = prop as
		| Record<string, boolean | number | string | undefined>
		| undefined
	return createEnv({
		onValidationError: (error: ZodError) => {
			console.error(
				'❌ Invalid environment variables:',
				error.flatten().fieldErrors,
			)
			process.exit(1)
		},
		// onInvalidAccess: (error: ZodError) => { },
		skipValidation: isServer &&
			!!process.env.SKIP_ENV_VALIDATION &&
			process.env.SKIP_ENV_VALIDATION !== 'false' &&
			process.env.SKIP_ENV_VALIDATION !== '0',
		server: (isBuild
			? serverVariables.build
			: { ...serverVariables.runtime, ...serverVariables.build }) as
				& typeof serverVariables.build
				& typeof serverVariables.runtime,
		clientPrefix: 'PUBLIC_',
		client: {
			PUBLIC_API_URL: z.string(),
		},
		// check if fileEnv is empty, if so use import.meta.env, otherwise use fileEnv
		runtimeEnv: rEnv ? { ...rEnv, ...import.meta.env } : import.meta.env,
	})
}
```

**Reply by florian-lefebvre** [reply]
So maybe adapters can be given an API to override how public/private static/dynamic env are computed

**Reply by alexanderniebuhr** [reply]
I've just added some links, which might help understand some Cloudflare rules:

https://github.com/t3-oss/t3-env/issues/60
https://github.com/t3-oss/t3-env/issues/65
https://github.com/withastro/astro/issues/7137

**Reply by lilnasy** [reply]
Seconding this, I think it's best to not rely on `process.env` altogether. SvelteKit relies on a similar hack as us to make `$env/dynamic` automatically work _most of the time_, hiding the complexity.

`astro:env` could be a static only solution, and `Astro.env` the runtime only counterpart.

**Reply by florian-lefebvre** [reply]
Interesting! Then the only distinction between public/private would be the prefix. Sounds pretty good

**Reply by alexanderniebuhr** [reply]
Another piece of reference and context I don't want to loose:

https://discord.com/channels/830184174198718474/1202237340618067988

**Reply by florian-lefebvre** [reply]
Btw since you talked about build time variables, that's definitely interesting in CIs for platforms like Vercel and Netlify. For example, it's currently only possible on Netlify using this plugin (https://github.com/bencao/netlify-plugin-inline-functions-env) but it would be so much better to have this covered by Astro by default.

**Reply by alexanderniebuhr** [reply]
Docs from Cloudflare say not to overwrite/prefill `process.env` https://developers.cloudflare.com/workers/runtime-apis/nodejs/process/#processenv

**Reply by alexanderniebuhr** [reply]
Another update after consulting with Cloudflare. They suggest to stay away from process.env completely in Cloudflare. That strengthens the proposal from Arsh and really shows that adapters need to be able to configure this to the adapters hoster needs

**Comment by natemoo-re** [comment]
So many maintainers involved in this discussion! Just wanted to chime in that I really appreciate the amount of thought that's already gone into this and I'd be totally in favor of accepting this proposal already. We typically don't need all the details ironed out at this stage, just a general consensus that we want to official work on this problem.

Behind the scenes, I know there is absolutely a desire to move towards a system like this! Our current env handling is naive and comes with many footguns. It's a really important DX that's overlooked at the moment.

**Comment by bluwy** [comment]
Thanks for kicking this off! About the goals:

> * Be able to validate env vars (eg. `SECRET_TOKEN` must be at least 64 long)
> * Be able to transform env vars (eg. `PORT` is always a number)

Initially, I quite like the idea of using `zod`. It feels very Astro 😄 But after reading others' comment on this, I agree too that maybe we could make this opt-in in the future. So that we avoid bundling `zod`, and potentially introducing a new file for env only. I think it would be best if one can setup env vars as quickly as possible.

---

Other than that, I wonder what we'll do with `import.meta.env.*`. In Vite by default , it's used as a public-static env var. Should we phase that out? It would be more consistent I think to converge on `astro:env`

**Reply by florian-lefebvre** [reply]
I think it could be great if `import.meta.env` was reserved for the built-in variables like `SSR` etc. I don't know how that works internally but I guess existing usage of `import.meta.env.MY_VAR` wouldn't break, just not recommended anymore

**Reply by florian-lefebvre** [reply]
> So that we avoid bundling zod, and potentially introducing a new file for env only.

That being said, I'm still convinced that adding something like 20+ lines of env config to the astro config makes it harder to navigate

**Reply by bluwy** [reply]
👍 Yeah I think it's not possible (or feasible) for us to completely disable `import.meta.env`, not recommending it anymore would be fine by me though.

**Comment by florian-lefebvre** [comment]
Closing as accepted
